### ISSUE TITLE
Kindle 2/DX crash because they lack a sysfs entry to read battery status

### ISSUE THREAD

**Issue by Yaanu** [koreader/koreader#5159]
* KOReader version: v2019.07
* Device: Kindle 2 International (B003, K2I)

#### Issue

I've installed KOReader to my K2I via KUAL, but when I attempt to launch it the program crashes before booting into anything. According to the crash log, it has something to do with the inability to register some fonts upon loading the QuickStart help page. There appear to be a number of other potential issues, but loading the fonts is where the program crashes and closes out.

#### Steps to reproduce

1. Get a Kindle 2 International
2. Install KUAL
3. Install KOReader
4. Launch KOReader
5. Check the crash.log

##### `crash.log` (if applicable)
[crash.log](https://github.com/koreader/koreader/files/3454438/crash.log)

**Comment by NiLuJe** [comment]
Nope, the relevant bit is actually in the stack trace, it actually crashes because The Kindle 2 & DX don't provide a `batt_capacity_file`.

I'm slightly puzzled as to why this was never an issue before, though (I'm fairly sure I tested some stuff on a K2i somewhat recently).

**Comment by NiLuJe** [comment]
.. Or not :D.
![lol](https://user-images.githubusercontent.com/111974/62262929-bdebb000-b41a-11e9-90f1-20e68bcc0deb.png)

Must have been thinking of the K3 ;p.

**Comment by NiLuJe** [comment]
TL;DR: There's no sysfs file *at all* on those Kindles.

We'll have to parse gasgauge-info -c's output (because it's bitbang'ing over i2c to read the value, so, no thank you to the DIY approach ;p).

**Comment by NiLuJe** [comment]
Well, at least now it boots and mostly works.

That said, everything is going to be painfully slow and cumbersome, because, well, it's a K2.

Also, SQLite is apparently broken for some reason (I'm going to guess the tremendously old Kernel, and the weird and broke-ass fuse proxy layer on top of vfat).

```
08/01/19-06:38:13 ERROR Failed to initialize coverbrowser plugin:  common/lua-ljsqlite3/init.lua:28: ljsqlite3[ioerr] disk I/O error
08/01/19-06:38:14 ERROR Failed to initialize statistics plugin:  common/lua-ljsqlite3/init.lua:28: ljsqlite3[ioerr] disk I/O error
```

**Comment by NiLuJe** [comment]
Okay, I lied.

```diff
diff --git a/thirdparty/lua-ljsqlite3/init.lua b/thirdparty/lua-ljsqlite3/init.lua
index 58d41b5..773256d 100644
--- a/thirdparty/lua-ljsqlite3/init.lua
+++ b/thirdparty/lua-ljsqlite3/init.lua
@@ -24,8 +24,8 @@ local xsys = require "xsys"
 
 local split, trim = xsys.string.split, xsys.string.trim
 
-local function err(code, msg)
-  error("ljsqlite3["..code.."] "..msg)
+local function err(code, extcode, msg)
+  error("ljsqlite3[" .. code .. ":" .. extcode .. "] "..msg)
 end
 
 -- Codes -----------------------------------------------------------------------
@@ -34,8 +34,9 @@ local codes = {
   [0] = "OK", "ERROR", "INTERNAL", "PERM", "ABORT", "BUSY", "LOCKED", "NOMEM",
   "READONLY", "INTERRUPT", "IOERR", "CORRUPT", "NOTFOUND", "FULL", "CANTOPEN",
   "PROTOCOL", "EMPTY", "SCHEMA", "TOOBIG", "CONSTRAINT", "MISMATCH", "MISUSE",
-  "NOLFS", "AUTH", "FORMAT", "RANGE", "NOTADB", [100] = "ROW", [101] = "DONE"
-} -- From 0 to 26.
+  "NOLFS", "AUTH", "FORMAT", "RANGE", "NOTADB", "NOTICE", "WARNING",
+  [100] = "ROW", [101] = "DONE"
+} -- From 0 to 28.
 
 do
   local types = { "INTEGER", "FLOAT", "TEXT", "BLOB", "NULL" } -- From 1 to 5.
@@ -64,7 +65,7 @@ do
 
   local t = sqlconstants
   local pre = "static const int32_t SQLITE_"
-  for i=0,26    do t[#t+1] = pre..codes[i].."="..i..";\n" end
+  for i=0,28    do t[#t+1] = pre..codes[i].."="..i..";\n" end
   for i=100,101 do t[#t+1] = pre..codes[i].."="..i..";\n" end
   for i=1,5     do t[#t+1] = pre..types[i].."="..i..";\n" end
   pre = pre.."OPEN_"
@@ -87,6 +88,8 @@ typedef struct Mem sqlite3_value;
 
 // Get informative error message.
 const char *sqlite3_errmsg(sqlite3*);
+int sqlite3_extended_errcode(sqlite3*);
+const char *sqlite3_errstr(int);
 
 // Connection.
 int sqlite3_open_v2(const char *filename, sqlite3 **ppDb, int flags,
@@ -189,13 +192,13 @@ conn_mt.__index = conn_mt
 
 -- Helper function to get error msg and code from sqlite.
 local function codemsg(pconn, code)
-  return codes[code]:lower(), ffi.string(sql.sqlite3_errmsg(pconn))
+  return codes[code]:lower(), sql.sqlite3_extended_errcode(pconn), ffi.string(sql.sqlite3_errmsg(pconn))
 end
 
 -- Throw error for a given connection.
 local function E_conn(pconn, code)
-  local code, msg = codemsg(pconn, code)
-  err(code, msg)
+  local code, extcode, msg = codemsg(pconn, code)
+  err(code, extcode, msg)
 end
 
 -- Test code is OK or throw error for a given connection.
@@ -207,7 +210,7 @@ end
 
 local function T_open(x)
   if x._closed then
-    err("misuse", "object is closed")
+    err("misuse", "n/a", "object is closed")
   end
 end
 
@@ -228,7 +231,7 @@ return function(stmt_or_value <opt_i>)
   elseif t == sql.SQLITE_NULL then
     return nil
   else
-    err("constraint", "unexpected SQLite3 type")
+    err("constraint", "n/a", "unexpected SQLite3 type")
   end
 end
 ]=]
@@ -250,7 +253,7 @@ return function(stmt_or_value, v <opt_i>)
   elseif t == "nil" then
     return sql.sqlite3_<variant>_null(stmt_or_value <opt_i>)
   else
-    err("constraint", "unexpected Lua type")
+    err("constraint", "n/a", "unexpected Lua type")
   end
 end
 ]=]
@@ -294,7 +297,7 @@ local function open(str, mode)
   mode = mode or "rwc"
   mode = open_modes[mode]
   if not mode then
-    err("constraint", "argument #2 to open must be ro, rw, or rwc")
+    err("constraint", "n/a", "argument #2 to open must be ro, rw, or rwc")
   end
   local aptr = ffi.new("sqlite3*[1]")
   -- Usually aptr is set even if error code, so conn always needs to be closed.
@@ -304,9 +307,9 @@ local function open(str, mode)
   connstmt[conn] = setmetatable({}, { __mode = "k" })
   conncb[conn] = { scalar = {}, step = {}, final = {} }
   if code ~= sql.SQLITE_OK then
-    local code, msg = codemsg(conn._ptr, code) -- Before closing!
+    local code, extcode, msg = codemsg(conn._ptr, code) -- Before closing!
     conn:close() -- Free resources, should not fail here in this case!
-    err(code, msg)
+    err(code, extcode, msg)
   end
   return conn
 end
@@ -358,7 +361,7 @@ function conn_mt:rowexec(command) T_open(self)
   local stmt = self:prepare(command)
   local res = stmt:_step()
   if stmt:_step() then
-    err("misuse", "multiple records returned, 1 expected")
+    err("misuse", "n/a", "multiple records returned, 1 expected")
   end
   stmt:close()
   if res then
@@ -552,7 +555,7 @@ function stmt_mt:resultset(get, maxrecords) T_open(self)
   get = get or "hik"
   maxrecords = maxrecords or math.huge
   if maxrecords < 1 then
-    err("constraint", "agument #1 to resultset must be >= 1")
+    err("constraint", "n/a", "argument #1 to resultset must be >= 1")
   end
   local hash, hasi, hask = get:find("h"), get:find("i"), get:find("k")
   local r, h = self:_step({}, {})
```

Yields:

```
common/lua-ljsqlite3/init.lua:28: ljsqlite3[ioerr:5386] disk I/O error
```

Meaning [SQLITE_IOERR_SHMMAP](https://www.sqlite.org/rescode.html#ioerr_shmmap).
Which means I was probably not so far off the mark, and the kernel doesn't support shared memory mappings.

**Comment by NiLuJe** [comment]
(mmap'ed IO is a requirement for WAL mode)
